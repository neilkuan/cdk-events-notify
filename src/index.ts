import * as path from 'path';
import * as events from '@aws-cdk/aws-events';
import * as target from '@aws-cdk/aws-events-targets';
import * as _lambda from '@aws-cdk/aws-lambda';
import * as logs from '@aws-cdk/aws-logs';
import * as cdk from '@aws-cdk/core';

export interface ISlackEventNotify {
  /**
   * slack Webhook Url for Lambda send message to slack.
   */
  readonly slackWebhookUrl: string;
  /**
   * slack Channel Name for Lambda send message to slack.
   */
  readonly slackChannelName: string;

}

export interface EventNotifyProps {
  /**
   * Line Notify Token for Lambda send notify permisson.
   */
  readonly lineNotifyToken?: string | undefined;

  /**
   * Notify target to Slack channel.
   */
  readonly slack?: ISlackEventNotify;
}

export class EventNotify extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props?: EventNotifyProps) {
    super(scope, id);

    if (!props?.lineNotifyToken && !props?.slack) {
      throw new Error('Please input lineNotifyToken or slack options');
    }

    const lambdaFun = new _lambda.Function(this, 'lambda_fun', {
      handler: 'linenotify.lambda_handler',
      code: _lambda.Code.fromAsset(path.join(__dirname, '../function') ),
      runtime: _lambda.Runtime.PYTHON_3_8,
      timeout: cdk.Duration.minutes(3),
      logRetention: logs.RetentionDays.THREE_DAYS,
      environment: {
        LINE_NOTIFY_TOKEN: props?.lineNotifyToken ? props?.lineNotifyToken : 'none',
        SLACK_WEBHOOK_URL: props?.slack?.slackWebhookUrl ? props?.slack?.slackWebhookUrl : 'none',
        SLACK_CHANNEL_NAME: props?.slack?.slackChannelName ? props?.slack?.slackChannelName : 'none',
      },
    });

    const eventNotifyTarget = new target.LambdaFunction(lambdaFun);
    new events.Rule(this, 'EventBusConsoleLoginEvent', {
      ruleName: 'EventBusConsoleLoginEvent',
      description: 'Console login in specific region',
      targets: [eventNotifyTarget],
      eventPattern: {
        detailType: ['AWS Console Sign In via CloudTrail'],
      },
    });
  }
}