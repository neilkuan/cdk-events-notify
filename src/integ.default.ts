import * as cdk from 'aws-cdk-lib';
import { EventNotify } from './';


const app = new cdk.App();

[
  'us-east-1',
  'ap-northeast-1',
].forEach(region => {
  const stack = new cdk.Stack(app, `integ-stack-${region}`, {
    env: {
      region: region,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    },
  });
  new EventNotify(stack, 'LineslackEventNotify', {
    lineNotifyToken: process.env.LINE_NOTIFY_TOKEN,
    slack: {
      slackChannelName: `${process.env.SLACK_CHANNEL_NAME}`,
      slackWebhookUrl: `${process.env.SLACK_WEBHOOK_URL}`,
    },
  });
});


