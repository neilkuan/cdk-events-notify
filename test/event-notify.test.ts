import '@aws-cdk/assert/jest';
import * as cdk from '@aws-cdk/core';
import { EventNotify } from '../src/index';

test('line and slack', () => {
  const mockApp = new cdk.App();
  const stack = new cdk.Stack(mockApp, 'testing-stack');
  new EventNotify(stack, 'testing', {
    lineNotifyToken: 'mock',
    slack: {
      slackChannelName: 'mock',
      slackWebhookUrl: 'mock',
    },
  });
  expect(stack).toHaveResourceLike('AWS::Lambda::Function', {
    Code: {
      S3Bucket: {
        Ref: 'AssetParameters87b9cfa46edf9eec80d013291c5bdae277c3f45d382ff3919f8b76335730f435S3Bucket642DCB0A',
      },
      S3Key: {
        'Fn::Join': [
          '',
          [
            {
              'Fn::Select': [
                0,
                {
                  'Fn::Split': [
                    '||',
                    {
                      Ref: 'AssetParameters87b9cfa46edf9eec80d013291c5bdae277c3f45d382ff3919f8b76335730f435S3VersionKey65CD780B',
                    },
                  ],
                },
              ],
            },
            {
              'Fn::Select': [
                1,
                {
                  'Fn::Split': [
                    '||',
                    {
                      Ref: 'AssetParameters87b9cfa46edf9eec80d013291c5bdae277c3f45d382ff3919f8b76335730f435S3VersionKey65CD780B',
                    },
                  ],
                },
              ],
            },
          ],
        ],
      },
    },
    Role: {
      'Fn::GetAtt': [
        'testinglambdafunServiceRole62CC8ACF',
        'Arn',
      ],
    },
    Environment: {
      Variables: {
        LINE_NOTIFY_TOKEN: 'mock',
        SLACK_WEBHOOK_URL: 'mock',
        SLACK_CHANNEL_NAME: 'mock',
      },
    },
    Handler: 'linenotify.lambda_handler',
    Runtime: 'python3.8',
    Timeout: 180,
  });
});

test('slack only', () => {
  const mockApp = new cdk.App();
  const stack = new cdk.Stack(mockApp, 'testing-stack');
  new EventNotify(stack, 'testing', {
    slack: {
      slackChannelName: 'mock',
      slackWebhookUrl: 'mock',
    },
  });
  expect(stack).toHaveResourceLike('AWS::Lambda::Function', {
    Code: {
      S3Bucket: {
        Ref: 'AssetParameters87b9cfa46edf9eec80d013291c5bdae277c3f45d382ff3919f8b76335730f435S3Bucket642DCB0A',
      },
      S3Key: {
        'Fn::Join': [
          '',
          [
            {
              'Fn::Select': [
                0,
                {
                  'Fn::Split': [
                    '||',
                    {
                      Ref: 'AssetParameters87b9cfa46edf9eec80d013291c5bdae277c3f45d382ff3919f8b76335730f435S3VersionKey65CD780B',
                    },
                  ],
                },
              ],
            },
            {
              'Fn::Select': [
                1,
                {
                  'Fn::Split': [
                    '||',
                    {
                      Ref: 'AssetParameters87b9cfa46edf9eec80d013291c5bdae277c3f45d382ff3919f8b76335730f435S3VersionKey65CD780B',
                    },
                  ],
                },
              ],
            },
          ],
        ],
      },
    },
    Role: {
      'Fn::GetAtt': [
        'testinglambdafunServiceRole62CC8ACF',
        'Arn',
      ],
    },
    Environment: {
      Variables: {
        LINE_NOTIFY_TOKEN: 'none',
        SLACK_WEBHOOK_URL: 'mock',
        SLACK_CHANNEL_NAME: 'mock',
      },
    },
    Handler: 'linenotify.lambda_handler',
    Runtime: 'python3.8',
    Timeout: 180,
  });
});

test('line only', () => {
  const mockApp = new cdk.App();
  const stack = new cdk.Stack(mockApp, 'testing-stack');
  new EventNotify(stack, 'testing', {
    lineNotifyToken: 'mock',
  });
  expect(stack).toHaveResourceLike('AWS::Lambda::Function', {
    Code: {
      S3Bucket: {
        Ref: 'AssetParameters87b9cfa46edf9eec80d013291c5bdae277c3f45d382ff3919f8b76335730f435S3Bucket642DCB0A',
      },
      S3Key: {
        'Fn::Join': [
          '',
          [
            {
              'Fn::Select': [
                0,
                {
                  'Fn::Split': [
                    '||',
                    {
                      Ref: 'AssetParameters87b9cfa46edf9eec80d013291c5bdae277c3f45d382ff3919f8b76335730f435S3VersionKey65CD780B',
                    },
                  ],
                },
              ],
            },
            {
              'Fn::Select': [
                1,
                {
                  'Fn::Split': [
                    '||',
                    {
                      Ref: 'AssetParameters87b9cfa46edf9eec80d013291c5bdae277c3f45d382ff3919f8b76335730f435S3VersionKey65CD780B',
                    },
                  ],
                },
              ],
            },
          ],
        ],
      },
    },
    Role: {
      'Fn::GetAtt': [
        'testinglambdafunServiceRole62CC8ACF',
        'Arn',
      ],
    },
    Environment: {
      Variables: {
        LINE_NOTIFY_TOKEN: 'mock',
        SLACK_WEBHOOK_URL: 'none',
        SLACK_CHANNEL_NAME: 'none',
      },
    },
    Handler: 'linenotify.lambda_handler',
    Runtime: 'python3.8',
    Timeout: 180,
  });
});