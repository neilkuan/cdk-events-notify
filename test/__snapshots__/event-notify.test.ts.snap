// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integ snapshot validation 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters5060827a373976ce216bd21151047cfd6e0ffc961f7c120448a6f9fecc2a3244ArtifactHash028DE8F3": Object {
      "Description": "Artifact hash for asset \\"5060827a373976ce216bd21151047cfd6e0ffc961f7c120448a6f9fecc2a3244\\"",
      "Type": "String",
    },
    "AssetParameters5060827a373976ce216bd21151047cfd6e0ffc961f7c120448a6f9fecc2a3244S3BucketDF38F68F": Object {
      "Description": "S3 bucket for asset \\"5060827a373976ce216bd21151047cfd6e0ffc961f7c120448a6f9fecc2a3244\\"",
      "Type": "String",
    },
    "AssetParameters5060827a373976ce216bd21151047cfd6e0ffc961f7c120448a6f9fecc2a3244S3VersionKeyC6AB4B1A": Object {
      "Description": "S3 key for asset version \\"5060827a373976ce216bd21151047cfd6e0ffc961f7c120448a6f9fecc2a3244\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24ArtifactHashBA91B77F": Object {
      "Description": "Artifact hash for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5": Object {
      "Description": "S3 bucket for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861": Object {
      "Description": "S3 key for asset version \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "LineEeventNotifyEventBusConsoleLoginEvent89BFF1A3": Object {
      "Properties": Object {
        "Description": "Console login in specific region",
        "EventPattern": Object {
          "detail-type": Array [
            "AWS Console Sign In via CloudTrail",
          ],
        },
        "Name": "EventBusConsoleLoginEvent",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "LineEeventNotifylambdafun5C90B316",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "LineEeventNotifyEventBusConsoleLoginEventAllowEventRuleintegstackLineEeventNotifylambdafun2178BEF44BF1F3B3": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "LineEeventNotifylambdafun5C90B316",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "LineEeventNotifyEventBusConsoleLoginEvent89BFF1A3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "LineEeventNotifylambdafun5C90B316": Object {
      "DependsOn": Array [
        "LineEeventNotifylambdafunServiceRoleC923ABE4",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters5060827a373976ce216bd21151047cfd6e0ffc961f7c120448a6f9fecc2a3244S3BucketDF38F68F",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5060827a373976ce216bd21151047cfd6e0ffc961f7c120448a6f9fecc2a3244S3VersionKeyC6AB4B1A",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5060827a373976ce216bd21151047cfd6e0ffc961f7c120448a6f9fecc2a3244S3VersionKeyC6AB4B1A",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "LINE_NOTIFY_TOKEN": "mock",
          },
        },
        "Handler": "linenotify.lambda_handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LineEeventNotifylambdafunServiceRoleC923ABE4",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 180,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LineEeventNotifylambdafunLogRetentionD82988F8": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "LineEeventNotifylambdafun5C90B316",
              },
            ],
          ],
        },
        "RetentionInDays": 3,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LineEeventNotifylambdafunServiceRoleC923ABE4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
